From b96ec32bd9a3a3f7e70b548542097cb8a536094d Mon Sep 17 00:00:00 2001
From: Peter Swain <pswain@cavium.com>
Date: Tue, 1 Jan 2019 23:37:43 -0800
Subject: [PATCH 231/767] mmc: cavium: finish_dma_single() should
 teardown/unmap

commit e5bd68bd1a00ebe5faff4afd1d7cca8142418fa1 from
git@git.assembla.com:cavium/WindRiver.linux.git

Since its introduction, finish_dma_single() has left the
DMA mapping in place, slowly depleting the iommu stream's
address space, and stalling transfers when the pool is empty

Change-Id: I6371f29682cd681d9c86b93de0e9db92c6f0d22d
Signed-off-by: Peter Swain <pswain@cavium.com>
Reviewed-on: https://sj1git1.cavium.com/2240
Tested-by: sa_ip-sw-jenkins
Reviewed-by: Chandrakala Chavva <Chandrakala.Chavva@cavium.com>
Reviewed-on: https://sj1git1.cavium.com/8541
Tested-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
Reviewed-by: Chandrakala Chavva <cchavva@marvell.com>
[Kevin: Drop the dma_unmap_sg() since it was already applied]
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 drivers/mmc/host/cavium.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/drivers/mmc/host/cavium.c b/drivers/mmc/host/cavium.c
index e02fd0119d0a..f308decf6410 100644
--- a/drivers/mmc/host/cavium.c
+++ b/drivers/mmc/host/cavium.c
@@ -495,6 +495,9 @@ static int finish_dma_single(struct cvm_mmc_host *host, struct mmc_data *data)
 {
 	data->bytes_xfered = data->blocks * data->blksz;
 	data->error = 0;
+
+	/* Clear and disable FIFO */
+	writeq(BIT_ULL(16), host->dma_base + MIO_EMM_DMA_FIFO_CFG(host));
 	dma_unmap_sg(host->dev, data->sg, data->sg_len, get_dma_dir(data));
 	return 1;
 }
-- 
2.31.1

